### PR TITLE
Expose connectivity state

### PR DESCRIPTION
**Pull request by SebastianThiebaud**

### CODE REVIEW

File: Sources/gRPC/Channel.swift [hunk 1]
```diff
@@ -79,6 +82,131 @@ public class Channel {
...
+          run()
+        }
+      }
+    }
+
+    init(state: ConnectivityState, underlyingChannel: UnsafeMutableRawPointer, tryToConnect: Bool) {
+      self.state = state
+      self.underlyingChannel = underlyingChannel
+      self.tryToConnect = tryToConnect
+      self.queue = CompletionQueue(underlyingCompletionQueue: cgrpc_channel_completion_queue(underlyingChannel))
```
Comment (cvanderschuere) [inline Sources/gRPC/Channel.swift]: Give it a name?

File: Sources/gRPC/Channel.swift [hunk 2]
```diff
@@ -79,6 +82,131 @@ public class Channel {
...
+    }
+
+    private func run() {
+      DispatchQueue.global().async { [weak self] in
+        guard let `self` = self, let underlyingState = self.state.underlyingState else { return }
+
+        while self.polling == true {
```
Comment (cvanderschuere) [inline Sources/gRPC/Channel.swift]: Swift style is be explicit about this?

File: Sources/gRPC/Channel.swift [hunk 3]
```diff
@@ -79,6 +82,131 @@ public class Channel {
...
+          guard !self.callbacks.isEmpty || self.tryToConnect else {
```
Comment (cvanderschuere) [inline Sources/gRPC/Channel.swift]: What `|| self.tryToConnect`?

Comment (cvanderschuere) [reply Sources/gRPC/Channel.swift]: Reading the python implementation this should be `&& !self.tryToConnect`

File: Sources/gRPC/Channel.swift [hunk 4]
```diff
@@ -79,6 +82,131 @@ public class Channel {
...
+            self.polling = false
+            break
+          }
+
+          let deadline: TimeInterval = 0.2
+          cgrpc_channel_watch_connectivity_state(self.underlyingChannel, underlyingState, deadline, nil)
+          let event = self.queue.wait(timeout: deadline)
+
+          if event.success == 1 || self.tryToConnect {
+            let newState = ConnectivityState.connectivityState(cgrpc_channel_get_connectivity_state(self.underlyingChannel, self.tryToConnect ? 1 : 0))
```
Comment (cvanderschuere) [inline Sources/gRPC/Channel.swift]: You should set `tryToConnect` to false after we perform this check

File: Sources/gRPC/Channel.swift (hunk 3 above)
Comment (SebastianThiebaud) [reply Sources/gRPC/Channel.swift]: This is a guard statement, so I think I'm right here.

File: Sources/gRPC/Channel.swift (hunk 2 above)
Comment (SebastianThiebaud) [reply Sources/gRPC/Channel.swift]: `self` used to be a weak reference, cleaning it now.

File: Sources/gRPC/Channel.swift (hunk 1 above)
Comment (SebastianThiebaud) [reply Sources/gRPC/Channel.swift]: Done.

File: Sources/gRPC/Channel.swift (hunk 4 above)
Comment (SebastianThiebaud) [reply Sources/gRPC/Channel.swift]: Done.

File: Sources/gRPC/Channel.swift (hunk 3 above)
Comment (cvanderschuere) [reply Sources/gRPC/Channel.swift]: Yep you are

File: Sources/gRPC/Channel.swift [hunk 5]
```diff
@@ -123,11 +124,12 @@ private extension Channel {
       self.tryToConnect = tryToConnect
       self.queue = CompletionQueue(underlyingCompletionQueue: cgrpc_channel_completion_queue(underlyingChannel))
       self.queue.name = "Connectivity State"
+      self.lastState = ConnectivityState.connectivityState(cgrpc_channel_get_connectivity_state(self.underlyingChannel, 0))
```
Comment (cvanderschuere) [inline Sources/gRPC/Channel.swift]: There is going to be a race condition here with the passed in state and `state`

**Pull request merged**